### ISSUE TITLE
Access to model outputs inside LogitProcessor

### ISSUE THREAD

**Issue by AdityaMayukhSom** [huggingface/transformers#34265]
### Feature request

![image](https://github.com/user-attachments/assets/cb57d5dd-6502-4b31-a948-bb46e535fea5)

The LogitProcessor __call__ method currently has access to the input_ids and the logits generated. It would be really helpful if it has access to the model output for that iteration.

### Motivation

I am trying to implement something similar to [Get To The Point: Summarization with Pointer-Generator Networks](https://arxiv.org/abs/1704.04368) along with transformers, especially Llama models. I am required to add the attention output to the model's generated logit for pointing mechanism, but I do not have access to the attention value inside logit processor. As seen in the picture, if the standard model output object can be passes to the logit processor, then the user can extract required details from the output object and use it to further process the generated logits.

![image](https://github.com/user-attachments/assets/4103b2f8-a09c-4a2e-8916-9ebac8962dd1)

### Your contribution

I can implement the required changed and create a PR if allowed.

**Comment by zucchini-nlp** [comment]
cc @gante

**Comment by qgallouedec** [comment]
Hey @AdityaMayukhSom, sorry for the very slow reply, we're catching up on the backlog.

We'd rather not widen the `LogitsProcessor.__call__` contract for this: it is implemented by 34 built-in processors plus a lot of third-party ones, and "the model output for this iteration" has no clean meaning under beam search or assisted decoding, where one forward pass feeds several positions. You don't need it, though. The processor is called right after the forward pass, in the same loop iteration.

https://github.com/huggingface/transformers/blob/483b88066444bf67617efb7a26d6d37792b4b1d8/src/transformers/generation/utils.py#L3045-L3060

So a plain forward hook gives you exactly that step's output object:

```python
stash = {}
model.register_forward_hook(lambda m, args, out: stash.__setitem__("out", out))

class PointerProcessor(LogitsProcessor):
    def __call__(self, input_ids, scores):
        attentions = stash["out"].attentions      # this step's attentions
        hidden = stash["out"].hidden_states
        return scores

model.generate(**inputs, output_attentions=True, output_hidden_states=True,
               logits_processor=[PointerProcessor()])
```

Load the model with `attn_implementation="eager"` so attentions is actually populated. I double-checked this on current main: the processor sees the prefill attentions on the first step and the single-query attentions on each decode step, which is the per-iteration value you were after.

If you need more than a logits tweak, a pointer-generator head is a good fit for a custom generation method: https://huggingface.co/docs/transformers/generation_strategies#custom-generation-methods

Closing as won't-implement since no API change is needed here, but please re-open if the hook doesn't cover your setup.
